src/tech/yk_structure.md

```diff
@@ -0,0 +1,50 @@
+# The Structure of the JIT Runtime
+
+The `yk` repo's structure (and build system) is somewhat unusual as the code is
+spread over two workspaces. This is because we need to separately compile parts
+of the codebase with different configurations.
+
+Code that we expect to be able to hardware trace needs to be compiled with `-C
+tracer=hw`. Crucially, since hardware tracing relies on blocks not being
+reordered by LLVM, this flag disables optimisation. However, the JIT runtime
+will never be traced and therefore we can (and should) optimise this code.
+Further we want to be able to optimise the dependencies of the JIT runtime.
+
+To this end, we have two Rust workspaces in the `yk` repo: the "internal"
+(optimised) workspace, and the "external" (unoptimised) workspace. The external
+workspace then talks to the internal workspace via `extern` functions defined
+in the `ykshim` crate.
+
+```
```

Comment:
The vertical line here is a bit confusing. I wonder if we should add a top legend along the lines of:

```
        external workspace   |  internal workspace
----------------------------------------------------------------
```

and remove the big vertical line?

Reply:
would a more dashed line help? Using the ellipsis character? `⋮`

Reply:
I don't think it would help, because it would still look like an edge in the graph.

Reply:
I've pushed a commit with your suggestion, which looks like this:
```
       External Workspace             |       Internal Workspace
--------------------------------------------------------------------

interpreter --\
              |
              +---> ykshim_client --------> ykshim --> JIT runtime
              |
tests  -------/
```

but may I also suggest:
```
       External Workspace             .       Internal Workspace
                                      .
                                      .
interpreter --\                       .
              |                       .
              +---> ykshim_client --------> ykshim --> JIT runtime
              |                       .
tests  -------/                       .
```

Which do you prefer?

Reply:
First one for me.

Reply:
How about:
```
          External Workspace                Internal Workspace

+------------------------------------+ +-----------------------------+
|                                    | |                             |
| Interpreter +                      | |                             |
|             |                      | |                             |
|             +----> ykshim_client +----> yk_shim +----> JIT runtime |
|             |                      | |                             |
|       Tests +                      | |                             |
|                                    | |                             |
+------------------------------------+ +-----------------------------+
```

Reply:
@ltratt did you like this? If you do, I can do another PR.

Reply:
Works for me!